20069

MarcelEdward opened the pull request:
In Application.cpp the 3Dconnexion.h file in inckuded and the init function is called initializing the device.

The native support for the 3Dconnexion device is added in the 3Dconnexion.cpp en 3Dconnexion.h files.

There is a connexionData class for the input mapper.  There is a connexionClient class wich is different for osx and windows. Windows has got a separate params class.
The osx connexionClient is more or less the official api implementation .
The win connexionClient ia  the unofficial way to implement. It reads the raw input usb data, The official api does not seem to work under windows with c++ and qt.

hifi-gustavo commented:
An admin must verify this patch for the pull request build to fire. Type "ok to test" to begin testing this pull request or "add to whitelist" to test this and all future pull requests from this collaborator.

MarcelEdward commented:
I have put sdks, documentation and samples from 3Dconnexion on http://meverhagen.nl/DocumentationSamples.zip

ZappoMan commented:
ok to test

hifi-gustavo commented:
Build console output for commit: 31a5e94885e48632b7becd2b0e909ef32abbd5d9 available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

MarcelEdward commented:
The apple build needs the 3Dconnexion driver to be installed from http://www.3dconnexion.eu/service/drivers.html

hifi-gustavo commented:
Build console output for commit: 47780740b7a13e4dfa3fa83f1201380514fa202b available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

AndrewMeadows commented on interface/external/connexionclient/Inc/I3dMouseParams.h:
```diff
@@ -0,0 +1,84 @@
+
+protected:
+
+virtual ~I3dMouseNavigation(){}
```
These function declarations (and others in this PR not itemized) should be indented -- although it the coding standard doesn't mention this specifically, it is the style shown in section 2.1.3 and is used throughout the codebase.

The whitespace formatting of the functions names so that they line up should be removed as per coding standard section 4.2.7.

https://docs.highfidelity.com/v1.0/docs/coding-standard

AndrewMeadows commented on interface/external/connexionclient/readme.txt:
```diff
@@ -0,0 +1,91 @@
+
+#endif
+
+
```
Is this a readme.txt file or a header file?

AndrewMeadows commented on interface/src/devices/3Dconnexion.h:
```diff
@@ -0,0 +1,226 @@
+
+#endif /* __APPLE__ */
+
+#endif /* defined(hifi_ConnexionClient_h) */
```
This file appears to use tabs for indentation.  It should use four spaces per indentation level instead (coding standard section 2.1.5).

AndrewMeadows replied:
There is much OS dependent preprocessor stuff going on here.  Not a blocker but: I wonder if there is a way to organize it better so that things are more readable.

AndrewMeadows commented on interface/src/devices/3Dconnexion.cpp:
```diff
@@ -0,0 +1,1016 @@
+#endif /* __APPLE__ */
+
+
+
```
This file appears to use tabs for indentation. It should use four spaces per indentation level instead (coding standard section 2.1.5).

AndrewMeadows commented on interface/src/devices/3Dconnexion.cpp:
```diff
@@ -0,0 +1,1016 @@
+//
+//  Created by MarcelEdward Verhagen on 09-06-15.
+//
+//
```
Needs Apache-compatible license statement here.

AndrewMeadows commented on interface/external/connexionclient/Inc/I3dMouseParams.h:
```diff
@@ -0,0 +1,84 @@
+#ifndef I3D_MOUSE_PARAMS_H
+#define I3D_MOUSE_PARAMS_H
+
```
Needs Apache-compatible license statement and other info at the top of this file..

AndrewMeadows commented on interface/external/connexionclient/Inc/I3dMouseParams.h:
```diff
@@ -0,0 +1,84 @@
+	  kManualPivot = 0,
+	  kAutoPivot,
+	  kAutoPivotOverride
+   };
```
We don't use "enum EName" style.  It should be something like:

    enum Pivot {
        PIVOT_MANUAL = 0,
        PIVOT_AUTO,
        PIVOT_AUTO_OVERRIDE 
    };

Similarly for the enums below.  See coding standard section 1.2.15.

AndrewMeadows commented on interface/src/devices/3Dconnexion.cpp:
```diff
@@ -0,0 +1,1016 @@
+	if (buttonState == 0) {
+		_buttonPressedMap.erase(lastButtonState);
+	}
+	else {
```
Combine the two lines above to be:

    } else {

AndrewMeadows commented on interface/src/devices/3Dconnexion.cpp:
```diff
@@ -0,0 +1,1016 @@
+static void	DeviceRemovedHandler(unsigned int connection);
+static void	MessageHandler(unsigned int connection, unsigned int messageType, void *messageArgument);
+
+ConnexionData* connectdata = new ConnexionData();
```
Why is **connectData** newed up rather than being a proper static instance?  While it doesn't technically cause a memory leak (it is a valid singleton pointer until program shutdown) it will show up as a possible leak under code inspection tools such as valgrind.

Similarly for the code elsewhere in this PR:

    client = new ConnexionClient();

Is there a way to make that a proper instance rather than a newed instance?

AndrewMeadows commented on interface/src/devices/3Dconnexion.h:
```diff
@@ -0,0 +1,226 @@
+	void		On3dmouseInput();
+
+	class TInputData
+	{
```
Should use K&R formatting here (and other places) such that the opening brace is at the end of the line like this:

    class Foo {
    };

And also:

    if (logic) {
    } else {
    }

MarcelEdward commented:
Aw that is a lot, I should have studied the formatting rules better.

Replaced the tabs with four spaces
Updated the { } following the K&R formatting 
Added the apache license info
updated the indent of the functions and the whitespaces.
updated the readme.txt file, removed the copy of the header file
Replaced /* */ with //

Unfortunately 3Dconnexion has not got a cross platform sdk or ap. On osx, windows (and also on linux) it is a different kind of animal. So yes there is a lot of os dependent code in it. I haven't got an elegant solution for this. Although I minimized the #ifdef's, so i should be a bit better readable.

I will look into the newed instances and the enum style.

hyperlogic commented on interface/src/devices/3Dconnexion.h:
```diff
@@ -0,0 +1,226 @@
+	void registerToUserInputMapper(UserInputMapper& mapper);
+	void assignDefaultInputMapping(UserInputMapper& mapper);
+
+	void update(float deltaTime);
```
update() doesn't seem to be defined in the .cpp.  This would be a good place to check for an attached mouse and register it if one is found.  Right now it seems that the device is registered when you make a new client, and is never removed.

The logic flow should probably be something like:
- init a new client
- in update, check for a new device.  If one is found, register it.  If an attached device is no longer found, make sure to remove it from the UserInputMapper
- in destroy, also make sure to remove any registered devices

Also make sure you're calling focusOutEvent() from Application.

SamGondelman replied:
update() doesn't seem to be defined in the .cpp. This would be a good place to check for an attached mouse and register it if one is found. Right now it seems that the device is registered when you make a new client, and is never removed.

The logic flow should probably be something like:
- init a new client
- in update, check for a new device. If one is found, register it. If an attached device is no longer found, make sure to remove it from the UserInputMapper
- in destroy, also make sure to remove any registered devices

Also make sure you're calling focusOutEvent() from Application.

MarcelEdward replied:
I will look into the update() function.

The focusOutEvent() clears the mapper data, I have found the place to add in application. But when is this event triggered ?

hyperlogic replied:
It's triggered when the window loses focus and is supposed to make it so the controllers don't do anything until the window regains focus, although it might be a little broken right now.

SamGondelman replied:
It's triggered when the window loses focus and is supposed to make it so the controllers don't do anything until the window regains focus, although it might be a little broken right now.

hifi-gustavo commented:
Build console output for commit: 71569f408ec3d1c0b01042a9ede500db3c35108b available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

hifi-gustavo commented:
Build console output for commit: d680ebbc0d44344158398f30ac82c461a03bf674 available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

hifi-gustavo commented:
Build console output for commit: dadf804adc09510677390f6b05aea55980bc5d00 available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

hifi-gustavo commented:
Build console output for commit: 26cef6029115b83ca69548c17ac7b6dc783fc349 available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

MarcelEdward commented:
Ok, the build console output from hifi-gustavo is a bit strange. 
Updated the 4 files on github cause xcode gives some strange error when trying to commit and push the changes.

hifi-gustavo commented:
Build console output for commit: c5976f6d3a91455f52c240f15f8ce72b4b9495dc available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

hifi-gustavo commented:
Build console output for commit: 506f4d31118d7606f9ba0e2b59b558cbad473e39 available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

hifi-gustavo commented:
Build console output for commit: 792ed5c8c90ee8eb6964741ecb3f87e847482f4c available at: https://gist.github.com/2669ec9e1e9149b4040b<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

hifi-gustavo commented:
Build console output for commit: 90b48d40f677db3ba268969b429b806571dfa419 available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

hifi-gustavo commented:
Build console output for commit: 553182b6160718be8acfdfe1214f38f77912b12e available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

AndrewMeadows commented on interface/src/devices/3Dconnexion.cpp:
```diff
@@ -0,0 +1,1010 @@
+#endif // HAVE_CONNEXIONCLIENT
+
+
+
```
Indentation is incorrect for this file as well as some formatting (e.g. opening braces are on their own lines).

AndrewMeadows commented on interface/src/devices/3Dconnexion.h:
```diff
@@ -0,0 +1,237 @@
+        bool IsZero() {
+            return (0. == fAxes[0] && 0. == fAxes[1] && 0. == fAxes[2] &&
+                0. == fAxes[3] && 0. == fAxes[4] && 0. == fAxes[5]);
+        }
```
Should use **0.0f** instead of **0.**

hifi-gustavo commented:
Build console output for commit: c39833770d752b4ddade52eaf2b0b02f28189001 available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

hifi-gustavo commented:
Build console output for commit: 4f59425c83b9caecc48ad6091354a72ad1767602 available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

AndrewMeadows commented:
We're going to try to find our 3dConnexion mouse so we an test this.

SamGondelman commented:
On windows:
- Buttons don't seem to be released when you let go of them until the window loses focus.  Possible solution: clear the button map on every update, don't try to erase the buttons from the map when they are released.  Not sure how that would work with how you have the events set up.
- Rotation seems very slow, but I see they have a scale of .5, which seems reasonable.  Maybe the events aren't coming through frequently enough?

On mac:
- Seems to work pretty well.  I was seeing a lot of input lag, but that is likely due to another known bug.  On a debug build, though, it's unbearably laggy,

On both:
- Some possible issues around adding/removing from the UserInputMapper.  At one point, on mac, I had two clients showing up, and it seemed like on windows it would sometimes skip a device id?

Was there a reason you commented out the pitch controls?  Seems like they'd be useful.

MarcelEdward commented:
I will fix the windows button issue. 
Will set the rotation faster.

I also notices slow debugging, on my mac it runs about 25% slower, on windows even worse. Will pull the most recent version from master.

Will fix the issue with the input mapper adding/removing. The events most likely get called again before the input mapper has processed the adding/removing.

I have out commented the pitch controls since it did not work well with the device, the head also picked when just moving and was hard to get the head back in the correct position.

SamGondelman commented:
Maybe use a modifier for the pitch so it doesn't interfere with movement?  One of the buttons would suffice, but you'd have to change how zooming works.  Maybe something like:
- BUTTON_1 + rotation = pitch
- BUTTON_2 = zoom in
- BUTTON_1 + BUTTON_2 (which it looks like is BUTTON_3?) = zoom out

hifi-gustavo commented:
Build console output for commit: 226327e3703ef2502aed3a02e966b347973cbb3c available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

hifi-gustavo commented:
Build console output for commit: 728e0341ce66d31a8af0908922dd85ed9c78a5ce available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

hifi-gustavo commented:
Build console output for commit: 753d91de569f3e66e6a8317d100526cf3ac516a4 available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

MarcelEdward commented:
Added the 3Dconnexion menu option under Developer->Avatar which inits or removes the connexion events. Unplugging the device just removes the input handlers
Attemp to fix the adding and removing of the inputMapper

copied the pitch at button 1 down and the zooming with the other button and both buttons pressed. (It would be cool if the input mapper would support not only the up and down movement of the head, but also left and right)

Made the rotation on windows faster.

hifi-gustavo commented:
Build console output for commit: c46ef1fa730e95bedc5d6a9a9ae1ec7efbc484a9 available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

hifi-gustavo commented:
Build console output for commit: 1495be7f0d3f7878e12c97cbff3d699a6e2f8197 available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

hifi-gustavo commented:
Build console output for commit: 8cc696196f2a0238ec6dc4bfcbb60501e19fb922 available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

hifi-gustavo commented:
Build console output for commit: c365e581eeeb7760e69c1236210b6442301599bc available at: https://gist.github.com/9bc76e556e91c62135c1<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5261.dmg

MarcelEdward commented:
There was a merge conflict, so I tried to update 20069 to the latest. I am not sure why that gets into the pr and still has got some checks failed.

MarcelEdward commented:
I made a rebase on: https://github.com/highfidelity/hifi/pull/5351
